enhancements/monitoring/metrics-collection-profiles.md

```diff
@@ -0,0 +1,430 @@
+monitor per profile).
+
+```yaml
+---
+apiVersion: monitoring.coreos.com/v1
+kind: ServiceMonitor
+metadata:
+  labels:
+    k8s-app: telemeter-client
+    monitoring.openshift.io/collection-profile: full
+  name: telemeter-client
+  namespace: openshift-monitoring
+spec:
+  endpoints:
+  - bearerTokenFile: /var/run/secrets/kubernetes.io/serviceaccount/token
+    interval: 30s
+    port: https
+    scheme: https
+    tlsConfig:
+      <...>
+  jobLabel: k8s-app
+  selector:
+    matchLabels:
+      k8s-app: telemeter-client
+---
+apiVersion: monitoring.coreos.com/v1
+kind: ServiceMonitor
+metadata:
+  labels:
+    k8s-app: telemeter-client
+    monitoring.openshift.io/collection-profile: minimal
+  name: telemeter-client-minimal
+  namespace: openshift-monitoring
+spec:
+  endpoints:
+  - bearerTokenFile: /var/run/secrets/kubernetes.io/serviceaccount/token
+    interval: 30s
+    port: https
+    scheme: https
+    tlsConfig:
+      <...>
+  jobLabel: k8s-app
+  selector:
+    matchLabels:
+      k8s-app: telemeter-client
+  metricRelabelings:
+  - sourceLabels: [__name__]
+    action: keep
+    regex: "federate_samples|federate_filtered_samples"
+ ```
+
+Note: 
+- the `metricRelabeling` section only keeping two metrics, while the rest is
+dropped.
+- the metrics in the `keep` section were obtained with the help of a script that
+  parsed all Alerts, PrometheusRules and Console dashboards to determine what
+  metrics were actually being used.
+
+Finally, a team that adopts the metrics collection profile feature should also 
+add themselves and their component to the list under 
+[Infrastruture needed](#infrastructure-needed-optional)
+
+#### Variation [optional]
+
+NA
+
+### API Extensions
+
+- Addition of a new field to the `cluster-monitoring-config` ConfigMap
+
+```go
+type PrometheusK8sConfig struct {
+	// Defines the metrics collection profile that Prometheus uses to collect
+	// metrics from the platform components. Supported values are `full` or
+	// `minimal`. In the `full` profile (default), Prometheus collects all
+	// metrics that are exposed by the platform components. In the `minimal`
+	// profile, Prometheus only collects metrics necessary for the default
+	// platform alerts, recording rules, telemetry and console dashboards.
+	CollectionProfile CollectionProfile `json:"collectionProfile,omitempty"`
+}
+```
+
+### Implementation Details/Notes/Constraints
+
+Each OpenShift team that wants to adopt this feature will be responsible for
+providing the different monitors. This work is not trivial. Dependencies between
+operators and their metrics exist. This makes it difficult for developers to
+determine whether a given metric can be excluded from the `minimal` profile or
+not. To aid teams with this effort the monitoring team will provide:
+- a CLI tool that consumes a ServiceMonitor and a list of metrics and
+  generates the `minimal` ServiceMonitor for that operator.
```

Comment:
```suggestion
- a CLI tool that consumes a rule file and a list of metrics and
  generates the `minimal``metricRelabelings` for that operator.
```
(cc @jan--f)

Reply:
I'm confused about this comment, this bullet point refers to https://github.com/openshift/cluster-monitoring-operator/pull/2029 which the goal was always to provide a CLI tool to help the generation of ServiceMonitors

Reply:
Ah sorry, let me loop you in on this. In the latest biweekly 1-1 with @jan--f, I talked a bit about how working with `ServiceMonitor` may open us up to be responsible for fields within that schema that are not immediately related to the effort. A better approach would be to take in the rule file that the team relies on, and generate a `metricRelabelings` schema, which they can put in their `ServiceMonitor`, `PodMonitor`, or `Probe`, without us having to worry about the different kinds.

Reply:
Interesting, do have an example of which fields we could end up being responsible for? I'm not 100% sold on the idea since this would imply more manual actions for the OCP teams that would want to implement this feature. They would have to run this feature Copy&Paste and add an annotation. Furthermore, I'm not too concerned about the different kinds since AFAIK most projects just use `ServiceMonitor`. 
Regardless if this is really the direction we want I'm happy to comply.

Reply:
We discussed two things:

1. The tool adding the needed relabelings should not generate the full ServiceMonitor with all fields that are essentially unrelated to the collection profiles. So it could either produce only the relabel configs or (which Pranshuand I didn't discuss at the time) modify a given ServiceMonitor and add the needed relabelings. The latter might be more complicated since we'd have to pay attention to preexisting relabel configs, since they interact with each other and order is important. The idea was to have to tool only assemble the keep relabel action and leave it to developers to integrate that into their ServiceMonitor.
2. For ease of use it would be nice if this tool knew how to parse Prometheus rule files (or PrometheusRule CRs) and maybe also Grafana dashboards? This is probably easier to add to a script then to document for developers how to extract the needed metrics. A script could also query a running Prometheus instance for a lost of rules and extract metric names from that.

Reply:
@jan--f @rexagod https://github.com/openshift/cluster-monitoring-operator/pull/2053 let me know if this is what you had in mind. This basically works like so. 
```bash
→ go run hack/monitorgen/main.go --minimal-metrics-file=metrics.txt 
metricRelabelings:
- action: keep
  regex: (metric2|metric3|metric4)
  sourceLabels:
  - __name__
```
I still feel like parsing the SM and generating the full ServiceMonitor it's a friendlier workflow but I might be missing some context of the discussions

Reply:
I agree that generating a full SM is friendlier. However we'd have to address other existing relabel configs, besides existing drop actions (which might incidentally be totally valid to keep). E.g. an already existing keep action will drop everything not explicitly kept, so two keep actions will only retain the intersection.
Basically things are more complicated then telling users "give us a service monitor and we just add whats needed". By just providing our relabel config with the message "this is what we need, please integrate that" might be the clearer message.

Reply:
I see, yes that certainly makes sense 👍 but then https://github.com/openshift/cluster-monitoring-operator/pull/2053 works right?

Reply:
My 2 cents: I'm not sure that we already have monitors using the keep action (on the other side, we have some drop usage). The tool could be "smart" enough to flag the problem if it arises though.
I feel that providing the transformed monitors is the best approach because people aren't familiar with relabeling configs.

Reply:
Sure, producing a best guess modified SM is also an option and probably just works for the majority of users (since they don't have any preexisting relabel configs). We should however be cautious when there are existing relabel configs and flag the produced output for human verification.